src/core/init/duration_updater.ts

```diff
@@ -40,27 +62,139 @@ export default function DurationUpdater(
+      if (!isMediaSourceOpened) {
+        return EMPTY;
+      }
+      return whenSourceBuffersEndedUpdates$(mediaSource.sourceBuffers);
+    }),
+    take(1),
+    map(() : number | null => {
+      const maximumPosition = manifest.getMaximumPosition();
+      const { isLive } = manifest;
       // Some targets poorly support setting a very high number for durations.
       // Yet, in live contents, we would prefer setting a value as high as possible
       // to still be able to seek anywhere we want to (even ahead of the Manifest if
       // we want to). As such, we put it at a safe default value of 2^32 excepted
       // when the maximum position is already relatively close to that value, where
       // we authorize exceptionally going over it.
-      const newDuration = !manifest.isLive ? maximumPosition :
-                                             Math.max(Math.pow(2, 32),
-                                                      maximumPosition + YEAR_IN_SECONDS);
+      const newDuration = !isLive ? maximumPosition :
+                                    Math.max(Math.pow(2, 32),
+                                             maximumPosition + YEAR_IN_SECONDS);
 
+      if (mediaSource.duration >= newDuration ||
+          // Even if the MediaSource duration is different than the duration that
+          // we want to set now, the last duration we wanted to set may be the same,
+          // as the MediaSource duration may have been changed by the browser.
+          //
+          // In that case, we do not want to update it.
+          //
+          newDuration === lastSetDuration) {
+        return null;
+      }
       if (isNaN(mediaSource.duration) || !isFinite(mediaSource.duration) ||
-          Math.abs(mediaSource.duration - newDuration) > 0.01)
-      {
+          newDuration - mediaSource.duration > 0.01) {
         log.info("Init: Updating duration", newDuration);
         mediaSource.duration = newDuration;
+        return newDuration;
       }
+      return null;
     }),
-    ignoreElements());
+    catchError((err) => {
+      log.warn("Duration Updater: Can't update duration on the MediaSource.", err);
+      return observableOf(null);
+    })
+  );
+}
+
+/**
+ * Returns an Observable which will emit only when all the SourceBuffers ended
+ * all pending updates.
+ * @param {SourceBufferList} sourceBuffers
+ * @returns {Observable}
+ */
+function whenSourceBuffersEndedUpdates$(
+  sourceBuffers: SourceBufferList
+) : Observable<undefined> {
+  if (sourceBuffers.length === 0) {
+    return observableOf(undefined);
+  }
+  const updatingSourceBuffers$: Array<Observable<boolean>> = [];
+  for (let i = 0; i < sourceBuffers.length; i++) {
+    const sourceBuffer = sourceBuffers[i];
+    updatingSourceBuffers$.push(
+      observableMerge(
+        observableFromEvent(sourceBuffer, "updatestart").pipe(mapTo(true)),
+        observableFromEvent(sourceBuffer, "update").pipe(mapTo(false)),
```

Comment:
Should we prefer to use the `SourceBuffer.onupdateend` property instead?

Reply:
No!

In a previous review comment, we decided to choose `update` instead as it is closer to what we want (from https://www.w3.org/TR/media-source/#sourcebuffer-events):
  - `updating`: `updating` transition from `true` to `false` (this is what we want to catch)
  - `updateend`: The append or remove has ended (we don't care about that)

Reply:
We only care about `updating` and the MediaSource's readyState, as shown in the spec for setting the `duration` attribute: https://www.w3.org/TR/media-source/#attributes

This also corresponds to what we tested

Reply:
Thanks!